[core] platform agnostic executor via collective_rpc

youkaichao opened the pull request.

youkaichao committed:
remove all args

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
stash

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
Merge branch 'main' into remove_allargs

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
Merge branch 'main' into remove_allargs

youkaichao committed:
add rank during init

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
update env var

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
stash

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
stash

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
draft version

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix spec decode

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
rename

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
use collective_rpc

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
stash

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
gpu executor

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
gpu executor

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix kwargs

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
device name

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
tests

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix tests

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
refine

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
refine

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
refactor xpu

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
rename

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
rename

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
hpu

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
hpu

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
remove gpu executor

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
remove neuron executor

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
openvino

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
openvino

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
tpu

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
cpu

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
cuda

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
hpu and xpu

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
ray

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
_get_executor_cls

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
tod

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
Merge branch 'main' into remove_allargs

youkaichao committed:
ray

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
rename

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

MengqingCao commented on vllm/executor/ray_distributed_executor.py:
```diff
@@ -137,21 +145,24 @@ def _init_workers_ray(self, placement_group: "PlacementGroup",
                 num_gpus=num_gpus,
```
```suggestion
            from vllm.platforms import current_platform
            if current_platform.is_cuda_alike():
                worker = ray.remote(
                    num_cpus=0,
                    num_gpus=num_gpus,
                    scheduling_strategy=scheduling_strategy,
                    **ray_remote_kwargs,
                )(RayWorkerWrapper).remote(vllm_config=self.vllm_config)
            else:
                worker = ray.remote(
                    num_cpus=0,
                    num_gpus=0,
                    resources={current_platform.ray_device_key: num_gpus},
                    scheduling_strategy=scheduling_strategy,
                    **ray_remote_kwargs,
                )(RayWorkerWrapper).remote(vllm_config=self.vllm_config)
```

MengqingCao commented on vllm/executor/uniproc_executor.py:
```diff
@@ -0,0 +1,55 @@
+        return
+
+
+UniProcExecutorAsync = UniProcExecutor
```
QQ: Do we no longer need asynchronous executors? Will this affect performance?

MengqingCao replied:
> QQ: Do we no longer need asynchronous executors? Will this affect performance?

Never mind this, I notice the asynchronous methods are already maintained in `ExecutorBase`

MengqingCao commented on vllm/executor/uniproc_executor.py:
```diff
@@ -0,0 +1,55 @@
+                       kwargs: Optional[Dict] = None) -> List[Any]:
+        if kwargs is None:
+            kwargs = {}
+        answer = getattr(self.driver_worker, method)(*args, **kwargs)
```
Many devices have unsupported features, and they raise NotImplementedError with a certain error message before. I think it's better to use `try ... except NotImplementedError` here to keep the message.

youkaichao replied:
good point. added in [d26f3f0](https://github.com/vllm-project/vllm/pull/11256/commits/d26f3f0068f3774a6fbbe88a29a57e1f2ecf4b45)

youkaichao commented:
@tlrmchlsmth thanks for the review! I will fix these comments later. Do you have any concerns about this design of platform-agnostic executor?

tlrmchlsmth commented:
@youkaichao no high level concerns. Happy to see so much code deleted!

youkaichao commented:
For people reviewing this large PR (most change is about deleting and merging old code), the high-level idea can help you understand the PR better:

The core idea of this PR, is a `collective_rpc` interface. When we create the worker process, we will directly assign a rank to this worker. After setting up all workers, the executor always call `collective_rpc` on all workers.

If an rpc method requires different arguments for different ranks, the executor will pass in all arguments as a list, and each rank selects its own argument. Such an rpc call will have a special handling at the `WorkerWrapperBase` level.

In most cases, all ranks require the same arguments, and the executor just broadcasts the arguments to all ranks.

ruisearch42 commented:
> If an rpc method requires different arguments for different ranks, the executor will pass in all arguments as a list, and each rank selects its own argument. Such an rpc call will have a special handling at the `WorkerWrapperBase` level.

Overall looks good. Just to confirm, the special handling is not yet in `WorkerWrapperBase`? I don't seem to find it. Also have a question, if the arguments are large, would it be an extra overhead to send all arguments of all ranks to each worker?

JaheimLee commented:
I got an error using V1 with tp=2 when there were multiple requests at the same time.
```
ERROR 12-31 11:09:06 core.py:200] EngineCore hit an exception: Traceback (most recent call last):
ERROR 12-31 11:09:06 core.py:200]   File "/data/lijinghui/uv_projects/.venv/lib/python3.12/site-packages/vllm/v1/engine/core.py", line 193, in run_engine_core
ERROR 12-31 11:09:06 core.py:200]     engine_core.run_busy_loop()
ERROR 12-31 11:09:06 core.py:200]   File "/data/lijinghui/uv_projects/.venv/lib/python3.12/site-packages/vllm/v1/engine/core.py", line 232, in run_busy_loop
ERROR 12-31 11:09:06 core.py:200]     outputs = self.step()
ERROR 12-31 11:09:06 core.py:200]               ^^^^^^^^^^^
ERROR 12-31 11:09:06 core.py:200]   File "/data/lijinghui/uv_projects/.venv/lib/python3.12/site-packages/vllm/v1/engine/core.py", line 124, in step
ERROR 12-31 11:09:06 core.py:200]     output = self.model_executor.execute_model(scheduler_output)
ERROR 12-31 11:09:06 core.py:200]              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 12-31 11:09:06 core.py:200]   File "/data/lijinghui/uv_projects/.venv/lib/python3.12/site-packages/vllm/v1/executor/multiproc_executor.py", line 152, in execute_model
ERROR 12-31 11:09:06 core.py:200]     model_output = self.collective_rpc("execute_model",
ERROR 12-31 11:09:06 core.py:200]                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ERROR 12-31 11:09:06 core.py:200]   File "/data/lijinghui/uv_projects/.venv/lib/python3.12/site-packages/vllm/v1/executor/multiproc_executor.py", line 146, in collective_rpc
ERROR 12-31 11:09:06 core.py:200]     raise e
ERROR 12-31 11:09:06 core.py:200]   File "/data/lijinghui/uv_projects/.venv/lib/python3.12/site-packages/vllm/v1/executor/multiproc_executor.py", line 135, in collective_rpc
ERROR 12-31 11:09:06 core.py:200]     raise result
ERROR 12-31 11:09:06 core.py:200] AssertionError
ERROR 12-31 11:09:06 core.py:200] 
```
Is it related to this pr?

tlrmchlsmth commented on vllm/executor/executor_base.py:
```diff
@@ -40,6 +46,13 @@ def _init_executor(self) -> None:
         pass
 
     @abstractmethod
+    def collective_rpc(self,
+                       method: str,
+                       timeout: Optional[float] = None,
+                       args: Tuple = (),
+                       kwargs: Optional[Dict] = None) -> List[Any]:
+        pass
+
```
docstring?

youkaichao replied:
added now.

tlrmchlsmth commented on vllm/executor/executor_base.py:
```diff
@@ -137,3 +183,143 @@ async def check_health_async(self) -> None:
+        self,
+        execute_model_req: ExecuteModelRequest,
+    ) -> List[SamplerOutput]:
+        # TODO: unify into collective_rpc
```
What needs to happen for us to switch this to use collective_rpc?

youkaichao replied:
we need to design how to make it work for pipeline parallel.

tlrmchlsmth commented on vllm/executor/executor_base.py:
```diff
@@ -137,3 +183,143 @@ async def check_health_async(self) -> None:
         self.check_health()
+
+
+class DistributedExecutorBase(ExecutorBase):
```
Given that `DistributedExecutorBase` is used in V0 but not in V1, do you envision this going away when we transition to V1 entirely?

youkaichao replied:
yes.

tlrmchlsmth commented on vllm/worker/worker_base.py:
```diff
@@ -416,29 +458,39 @@ class WorkerWrapperBase:
+                from vllm.utils import init_cached_hf_modules
+                init_cached_hf_modules()
+
+    def adjust_rank(self, rank_mapping: Dict[int, int]) -> None:
+        if self.rank in rank_mapping:
+            self.rank = rank_mapping[self.rank]
```
Could you explain why you had to add this function?

This is potentially dangerous, since bugs will occur when `rank` is stored in multiple places and they aren't kept in sync. Is the `rank` not in the `ParallelConfig` anymore?

(Please add a docstring, including when it is OK to call this function.)

youkaichao replied:
explained in [7275d5a](https://github.com/vllm-project/vllm/pull/11256/commits/7275d5aa75a6c44eefc00ce4cf4d194662803139)

tlrmchlsmth reviewed (commented):
A few more review comments. Generally looks good, and just a few smaller comments/concerns

youkaichao commented on vllm/executor/ray_distributed_executor.py:
```diff
@@ -123,8 +129,10 @@ def _init_workers_ray(self, placement_group: "PlacementGroup",
+        worker_initial_ranks = []
         for bundle_id, bundle in enumerate(placement_group.bundle_specs):
-            if not bundle.get("GPU", 0):
+            if not bundle.get(current_platform.device_name, 0):
```
added in https://github.com/vllm-project/vllm/pull/11948

youkaichao committed:
revert device_name change

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/platforms/openvino.py:
```diff
@@ -23,7 +23,7 @@
 class OpenVinoPlatform(Platform):
     _enum = PlatformEnum.OPENVINO
-    device_name: str = "openvino"
+    device_name: str = "CPU" if "CPU" in envs.VLLM_OPENVINO_DEVICE else "GPU"
```
it is not necessary after https://github.com/vllm-project/vllm/pull/11948 now. sorry for the confusion.

youkaichao committed:
Merge branch 'main' into executor_refactor

youkaichao committed:
fix linter

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
add RayWorkerMetaData

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
add comments

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
add comments

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix linter

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/executor/executor_base.py:
```diff
@@ -53,58 +66,94 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
-        raise NotImplementedError
+        # NOTE: This is logged in the executor because there can be >1 worker
+        # with other executors. We could log in the engine level, but work
+        # remains to abstract away the device for non-GPU configurations.
```
it should be `there can be >1 workers` . fixed now.

youkaichao committed:
format

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
format

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/executor/executor_base.py:
```diff
@@ -53,58 +66,94 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
+        sets = self.collective_rpc("list_prompt_adapters")
+        for s in sets:
+            assert s == sets[
+                0], "All workers should have the same prompt adapters."
```
fixed in [53cb755](https://github.com/vllm-project/vllm/pull/11256/commits/53cb755fea1d4c1b0a45fb0f923c222432777116)

youkaichao commented on vllm/worker/hpu_worker.py:
```diff
@@ -123,6 +124,70 @@ def init_device(self) -> None:
+            'VLLM_HPU_LOG_STEP_CPU_FALLBACKS_ALL', '0') != '0'
+        log_cpu_fallbacks = os.environ.get('VLLM_HPU_LOG_STEP_CPU_FALLBACKS',
+                                           '0') != '0' or log_cpu_fallbacks_all
+        if log_graph_compilation or log_cpu_fallbacks:
```
yes, and i think it should be fine, i don't want to leak this abstraction into executor.

youkaichao commented on vllm/executor/executor_base.py:
```diff
@@ -137,3 +183,143 @@ async def check_health_async(self) -> None:
+        # This is non-None when the execute model loop is running
+        # in the parallel workers. It's a coroutine in the AsyncLLMEngine case.
+        self.parallel_worker_tasks: Optional[Union[Any, Awaitable[Any]]] = None
+        # Updated by implementations that require additional args to be passed
+        # to the _run_workers execute_model call
+        self.extra_execute_model_run_workers_kwargs: Dict[str, Any] = {}
```
it seems this is dead code. let me delete it.

youkaichao replied:
fixed in [f56306b](https://github.com/vllm-project/vllm/pull/11256/commits/f56306b6f71db0e9a2b41e045a9c5193842a2636)

youkaichao committed:
remove extra_execute_model_run_workers_kwargs

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented:
> > If an rpc method requires different arguments for different ranks, the executor will pass in all arguments as a list, and each rank selects its own argument. Such an rpc call will have a special handling at the `WorkerWrapperBase` level.
> 
> Overall looks good. Just to confirm, the special handling is not yet in `WorkerWrapperBase`? I don't seem to find it. Also have a question, if the arguments are large, would it be an extra overhead to send all arguments of all ranks to each worker?

@ruisearch42 sorry for the late reply. you can check the `update_environment_variables` function, it has `envs = envs_list[self.rank]`, every rank selects its own argument.

> if the arguments are large, would it be an extra overhead to send all arguments of all ranks to each worker?

I expect most rpc calls have homogeneous arguments, and only a few rpc calls during initialization have heterogeneous arguments. therefore, in the common case, we only broadcast the shared argument to all workers.

youkaichao commented on vllm/worker/worker_base.py:
```diff
@@ -117,6 +116,49 @@ def list_loras(self) -> Set[int]:
         raise NotImplementedError
 
 
+class DelegateWorkerBase(WorkerBase):
+    """
+    A class that delegates all methods to another WorkerBase instance. This is
+    useful for creating a WorkerBase that wraps another WorkerBase instance,
+    e.g. speculative decoding.
+    """
```
it seems `spec_decode_worker.py` will use multi-step worker, and `medusa_worker.py` does not use worker wrapper base.

youkaichao committed:
add check_health

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/executor/ray_distributed_executor.py:
```diff
@@ -26,7 +26,7 @@
 
 
-class RayGPUExecutor(DistributedGPUExecutor):
+class RayDistributedExecutor(DistributedExecutorBase):
```
good point. fixed in [ebc6f22](https://github.com/vllm-project/vllm/pull/11256/commits/ebc6f2280cb6361590b3163f511cd10e81970874)

youkaichao committed:
add resources

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
format

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
format

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
Merge branch 'main' into executor_refactor

youkaichao committed:
Merge branch 'main' into executor_refactor

youkaichao committed:
use device_control_env_var

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented on vllm/executor/ray_distributed_executor.py:
```diff
@@ -227,20 +249,27 @@ def sort_by_driver_then_worker_ip(worker):
         # Set environment variables for the driver and workers.
         all_args_to_update_environment_variables = [({
-            "CUDA_VISIBLE_DEVICES":
+            f"{current_platform.device_name}_VISIBLE_DEVICES":
```
fixed in [c1698dd](https://github.com/vllm-project/vllm/pull/11256/commits/c1698dd0bdcbc919a864cad3e49f8e672942db13)

youkaichao committed:
call method directly

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
simplify code

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
simplify code

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented:
@tlrmchlsmth sorry for the long wait, I have addressed all the comments, and will work on passing all tests now.

let me know if you have further comments!

youkaichao committed:
fix tests

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix ray args

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix xpu

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix spec decode

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
use ray

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
Merge branch 'main' into executor_refactor

youkaichao committed:
fix hpu

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix xpu

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix xpu

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix neuron

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix neuron

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix neuron?

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao commented:
finally hardware tests passed.

youkaichao committed:
fix neuron

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix v1 code

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix ray

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix neuron

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix ray

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
unify initialize_ray_cluster

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix neuron

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
unify signature

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix format

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
lint

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
rename to init device

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
lint

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix v1 compatibility

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix neuron

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix neuron

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix neuron

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix neuron

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix ray executor for v1

Signed-off-by: youkaichao <youkaichao@gmail.com>

youkaichao committed:
fix shard state tests

Signed-off-by: youkaichao <youkaichao@gmail.com>

tlrmchlsmth reviewed (approved):
This looks good to me. Thanks for the great refactor!

@comaniac, @ruisearch42 should definitely check out the ray executor changes before landing this one

youkaichao commented on vllm/v1/worker/gpu_worker.py:
```diff
@@ -33,6 +33,7 @@ def __init__(
         local_rank: int,
         rank: int,
         distributed_init_method: str,
+        is_driver_worker: bool = False,
```
this is just for unifying the signature of workers, the variable is not used.

youkaichao commented:
per DM discussion, we can merge this pr first, and @ruisearch42 will follow on migrating the changes to v1.
